lib/hyrax/transactions/steps/set_collection_type_gid.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+module Hyrax
+  module Transactions
+    module Steps
+      ##
+      # A step that sets the `#collection_type_gid` in the change set.
+      #
+      # @since 2.4.0
+      class SetCollectionTypeGid
+        include Dry::Monads[:result]
+
+        ##
+        # @param [Hyrax::ChangeSet] change_set
+        # @param collection_type_gid [URI::GID] global id for the collection type
+        # @return [Dry::Monads::Result] `Failure` if there is no `collection_type_gid` or
+        #   it can't be set to the default for the input; `Success(input)`, otherwise.
+        def call(change_set, collection_type_gid: default_collection_type_gid)
```

Comment:
i'm wondering about the utility of this over treating `collection_type_gid` as a regular attribute? that is, doesn't `change_set.validate(collection_type_gid: some_gid)` work just as well?

i'm aware that at least some of the view forms and controllers don't include the `collection_type_gid` value in the collection params. the solution we came up with in Comet was to do `permitted = permitted.merge(collection_type_gid: params.require(:collection_type_gid))` in the controller. we felt like this user input interpretation was best handled directly in the controller. it doesn't seem any more complicated than adding step args, and removes a whole transaction step from the code base.

thoughts?

Reply:
i guess what i'm asking is: what is different about `collection_type_gid` vs any other attribute on this model/form? what's the reason for not setting it in the same way all the other attributes are set? likewise, why not handle defaults and validations in the same way defaults and validations are handled for other attributes (using `Valkyrie::ChangeSet`)?

i'd rather not confuse things with [comet's specifics](https://gitlab.com/surfliner/surfliner/-/blob/trunk/comet/app/controllers/hyrax/dashboard/collections_controller.rb#L19-41) (there's no collection transaction on comet's side, for no particularly good reason). the relevant code samples are included in my original comment.

i get the benefits of transactions in general, but in this case it's 35 extra lines of code and the effects i see are:

  - apply an attribute that would otherwise get applied with all the others in the `validate` step;
  - setup a default that will silently override any defaults set on the form object.

it just seems to complicate things to break this out into a separate step.

Reply:
`CollectionCreate` is modeled after the pattern in [WorkCreate](https://github.com/samvera/hyrax/blob/main/lib/hyrax/transactions/work_create.rb).  `WorkCreate` has steps [change_set.set_default_admin_set](https://github.com/samvera/hyrax/blob/main/lib/hyrax/transactions/steps/set_default_admin_set.rb), which sets the admin set to the default if one isn't assigned, and a second step [change_set.ensure_admin_set](https://github.com/samvera/hyrax/blob/main/lib/hyrax/transactions/steps/ensure_admin_set.rb), which verifies that the admin_set is set.  The transaction steps in `CollectionCreate` are consistent with those in `WorkCreate` in level of granularity.

Are you inclined to refactor the `WorkCreate` transactions?

Reply:
i think `change_set.set_default_admin_set` calls a service which has to create permission templates as well as assigning a variable.

if there's a way to neatly refactor those transaction steps to use form features, then yes, that seems very attractive to me. i'm not trying to have an argument about this though. i'm trying to provide a helpful review. if you don't want to change this, you can just say so.